Refactor rocoto task XML creation

WalterKolczynski-NOAA opened the pull request:
# Description
Refactors the rocoto task generation to be recursive. This will allow nested metatasks to loop over multiple variables, which is needed for GEFS product generation.

As part of this refactor, there is no longer separate arguments to designate metatasks. Instead, task dicts can include a nested 'task_dict' as well as a 'var_dict' containing the variables to loop over. The nested task dict can then either have another layer, or be the innermost task.

To accommodate the new recursive nature, some defaults that were previously defined in `create_wf_task()` had to be pushed down into the function that creates the innermost task. Also, former keywords have been absorbed by the task dict.

Refs #823
Refs #827

# Type of change
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Cycled test on Orion
- Forecast-only test on Orion
- GEFS test on Orion

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

WalterKolczynski-NOAA committed:
Refactor rocot task XML creation

Refactors the rocoto task generation to be recursive. This will allow
nested metatasks to loop over multiple variables, which is needed for
GEFS product generation.

As part of this refactor, there is no longer separate arguments to
designate metatasks. Instead, task dicts can include a nested 'task_dict'
as well as a 'var_dict' containing the variables to loop over. The nested
task dict can then either have another layer, or be the innermost task.

To accomodate the new recursive nature, some defaults that were previously
defined in `create_wf_task()` had to be pushed down into the function that
creates the innermost task. Also, former keywords have been absorbed by
the task dict.

Refs #823
Refs #827

WalterKolczynski-NOAA commented:
Not quite working yet, still need to make some tweaks.

WalterKolczynski-NOAA committed:
Add missing loop variables to task names

WalterKolczynski-NOAA committed:
Fix envars in new task definitions

WalterKolczynski-NOAA committed:
Fix typo in gempak meta jjob header call

WalterKolczynski-NOAA commented:
Okay, think I have the bug worked out.

WalterKolczynski-NOAA committed:
Fix a couple minor errors in new task dicts

HenryRWinterbottom reviewed (commented):
This LGTM. The docstrings are helpful and making the dictionaries more readable is easy on the eyes. This is beyond the scope of this PR, but with some time investment I think some of the redundant code blocks can be parsed out and stuffed in wrapper functions (e.g., decorators) to make this code even more readable. But, just my $0.02.

WalterKolczynski-NOAA commented:
Orion tests are randomly failing due to stalling on module load.

WalterKolczynski-NOAA committed:
Remove global-specific variable from rocoto.py

Returns `rocoto.py` to a workflow-agnostic state without needing
`cdump` or any rocoto entities. The middle-man `create_wf_task` in
task.py is also removed by explicitly setting several keys in the
task dict.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/ens_atmos_prod

WalterKolczynski-NOAA committed:
Define rotdir as Task field

aerorahul commented on workflow/rocoto/rocoto.py:
```diff
@@ -76,7 +123,7 @@ def create_task(task_dict: Dict[str, Any]) -> List[str]:
     ppn = resources_dict.get('ppn', 1)
     threads = resources_dict.get('threads', 1)
-    log = task_dict.get('log', 'demo.log')
+    log = task_dict.get('log', f'&ROTDIR;/logs/@Y@m@d@H/{taskname}.log')
```
I think this needs to be reverted.

WalterKolczynski-NOAA replied:
Oops. Fixed.

aerorahul reviewed (commented):
mostly looks good.  I think one tiny thing slipped through.

WalterKolczynski-NOAA committed:
Remove missed entity

WalterKolczynski-NOAA committed:
Fix GEFS dependencies

WalterKolczynski-NOAA committed:
Fix missing # for ecen task
